Olha hw4

olgastanitskaya opened the pull request:
Made home work #4.

olgastanitskaya committed:
#3 Created function crement and variable counter

olgastanitskaya committed:
#4 Created objects:
 - visitor,
 - snacks,
 - beverage,
 - bar.

 Also, implemented their methods.

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+            console.log("This product is already not available!!!")
+        }
+
+        return this.capacity > capacityDrink;
```
Тут больше имеет смысл >=, чем просто >

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+        } else if (beverage.type === "highAclogol") {
+            capacityDrink = 20;
+
+        } else if (beverage.type === "nonAlcohol") {
```
Такого типа напитков не существует в рамках этой задачи

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+            capacityDrink = 0;
+        }
+
+        if (this.capacity > capacityDrink) {
```
Почему не использовать функция canDink вместо копирования кода?

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+                    this.capacity -= capacityDrink;
+                    console.log(this.name + " drink " + beverage.name);
+                    break;
+                default:
```
Вывод сообщений  в консоль не предусмотрены задачей.
Не понятно зачем копировать тот же самый код в каждый case, если он вызывется безусловно - его можно выполнить 1 раз

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+        } else {
+            this.capacity += 5;
+        }
+        console.log(this.name + " ate " + snack.name);
```
вывод подобных сообщений не предусмотрен условиями задачи

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+    greet(visitor) {
+        console.log(`Hi, ${visitor.name}! Welcome to ${this.name}`)
+    },
+    expel(visitor) {
```
По условиям задачи функция не должна принимать аргументов

olgastanitskaya replied:
мне не понятно как я могу вывести имя посетителя? я могу сохранить имя в переменную в функции enter, но как получить это имя если функция ничего не принимает?

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+    },
+    decline(visitor) {
+        console.log(`SORRY ${visitor.name}! Today ${this.name} is full! See you later!`)
+    },
```
Все 3 функции выводят не корректные сообщения

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+    makeOrder(item) {
+        let result;
+        let count = 0;
+        this.drinks.forEach(index => {
```
1) я бы рассмотрел использование функций findIndex для поиска нужного элемента
2) Очень странное название аргумента - вводит в заблуждение, я бы рассмотрел вариант: **drink** или хотя бы **item**

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+        let result;
+        let count = 0;
+        this.drinks.forEach(index => {
+            result = index.name.includes(item);
```
не понятный мне и не правильный способ поиска. при таком способе вернет true для вариантов
```
"coca-cola".includes("cola");
"brandy-cola".includes("cola");
"colamiya".includes("cola");
```
Т.е. может быть совсем не то, что заказывал пользователь

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+                console.log(count);
+                let currentIndex = this.drinks.indexOf(index);
+                this.drinks.splice(currentIndex, 1);
+                return this.drinks;
```
Функция должна возвращать искомый напиток, а не список оставшихся напитков

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,277 @@
+                    this.snacks.splice(currentIndex, 1);
+                    return this.snacks;
+                }
+            });
```
Алгоритм поиска напитков и снеков выглядит довольно таки одинаково, можно вынести это в функцию вместо копирования кода

olgastanitskaya committed:
#4
1. Fixed condition ">" on ">=" in string "return this.capacity > capacityDrink;".
2. Delete extra type of beverage - "nonAlcohol" in the project.
3. Used function CanDrink that helps to go away copy/paste code.
4. Delete "console.log(this.name + " drink " + beverage.name);".
5. Delete "console.log(this.name + " ate " + snack.name);".
6. Change messages in functions: greet(),expel(),decline().
7. Used function "findIndex" and rename variable in this function.
8. Fixed algorithm search in the array without use function "includes()".
9.Function MakeOrder return found object
10.Added try..catch in methods "drink()" and "eat()"".
11.corrected spelling "Alcohol"

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -75,109 +72,93 @@ let visitor = {
-
-        } else {
-            this.puke();
+        }catch (e){
```
Не уверен, что try..catch тут к месту, для каких случаев это нужно?

rzubov replied:
try...catch тут не место, если есть необходимость проверить передано ли что-то в функцию или нет всегда это можно проверить в самом начале
```
if(!bevarage?.type){
   return;
}
```

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -75,109 +72,93 @@ let visitor = {
+                this.capacity += 5;
+            }
+        }catch (e){
+            console.log(`Этот товар уже недоступен!`);
```
try...catch тут не уместно

olgastanitskaya committed:
#4
1. Delete "try..catch" in methods: drink() and eat()
2. Change IF in "makeOrder()" and "canDrink()"
3. Update IF on SWITCH/CASE in method "eat()"
4. Created condition for method "eat()" that this.capacity can't be more 100

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,271 @@
+        } else if (beverage.type === "highAlcohol") {
+            capacityDrink = 20;
+        } else {
+            console.log("Этот товар уже недоступен!")
```
это не критично, но в этот else лишний

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,271 @@
+    age: 0,
+    capacity: 0,
+    canDrink(beverage) {
+        let capacityDrink;
```
можно инициализировать как 0

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,271 @@
+        this.capacity += 15;
+        if (this.capacity > 100) {
+            this.capacity = 100;
+        }
```
есть более короткий вариант
`this.capacity = Math.min(100, this.capacity);`

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,271 @@
+        }
+    },
+    eat: function (snack) {
+        try {
```
try..catch тут не нужно

rzubov reviewed (dismissed):
Поправь ошибки и можешь мерджишь

olgastanitskaya committed:
#4
1. Made check result function "makeOrder" in method "drink()"
2. Reduce code in method "puke()", used Math.min()
3. Initialization: let capacityDrink = 0;
4. Delete unnecessary ELSE in method "canDrink()"
5. Delete console.log in method "makeOrder()"

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+    canDrink(beverage) {
+        let capacityDrink = 0;
+
+        if (beverage.type === undefined) {
```
Это бесполезный if он ничего не делает

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+            return arrItem;
+        }
+
+        let Drink = findDrink(this.drinks);
```
Почему название переменной с большой буквы? Это запрещено!

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+    decline(visitor) {
+        console.log(`Извини ${visitor.name}! Сегодня ${this.name} переполнен! Попробуй позже!`);
+    },
+    makeOrder(item) {
```
Для чего эта функция менялась в последний раз? Тут же не было замечаний!

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+                return {
+                    Drink,
+                    Snack
+                }
```
Не понятно что тут вообще происходит

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+        }
+        if (this.capacity > 100) {
+            this.capacity = 100;
+        }
```
Тут так-же можно использовать Math.min

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+                this.capacity += 5;
+                break;
+            case undefined:
+                this.capacity += 0;
```
какой смысл добавлять 0?

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+    },
+    eat: function (snack) {
+
+        switch (snack.isEffective) {
```
Выглядит так, что тут всего 2 позиции - if подходит для этого лучше

rzubov commented on olgastanitskaya/hw4/bar.js:
```diff
@@ -0,0 +1,288 @@
+    puke() {
+
+        this.capacity += 15;
+        this.capacity = Math.min(100, this.capacity); //функция Math.min() выбираем минимальное число из переданных чисел
```
можно ее скоратить эти 2 строки в одну
` this.capacity = Math.min(100, this.capacity + 15)`

olgastanitskaya committed:
#4
1. delete unnecessary IF in method "canDrink()"
2. rename variables "Drink" and "Snack"
3. Change method "eat()"
4. Reduce code in method "puke()"

olgastanitskaya committed:
#4
1. delete unnecessary IF in method "canDrink()"
2. rename variables "Drink" and "Snack"
3. Change method "eat()"
4. Reduce code in method "puke()"

olgastanitskaya committed:
#4
1. Implemented function "Math.min();" in method "eat()""
